pipe.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"context"
+	"io"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	pb "github.com/kata-containers/agent/protocols/grpc"
+)
+
+type inPipe struct {
+	ctx         context.Context
+	agent       *shimAgent
+	containerId string
+	pid         uint32
+}
+
+func (p *inPipe) Write(data []byte) (n int, err error) {
+	resp, err := p.agent.WriteStdin(p.ctx, &pb.WriteStreamRequest{
+		ContainerId: p.containerId,
+		PID:         p.pid,
+		Data:        data})
+	if err != nil {
+		return 0, err
+	}
+
+	return int(resp.Len), nil
+}
+
+func (p *inPipe) Close() error {
+	_, err := p.agent.CloseStdin(p.ctx, &pb.CloseStdinRequest{
+		ContainerId: p.containerId,
+		PID:         p.pid})
+
+	return err
+}
+
+type outPipe struct {
+	ctx         context.Context
+	agent       *shimAgent
+	containerId string
+	pid         uint32
+}
+
+func (p *outPipe) Read(data []byte) (n int, err error) {
+	resp, err := p.agent.ReadStdout(p.ctx, &pb.ReadStreamRequest{
+		ContainerId: p.containerId,
+		PID:         p.pid,
+		Len:         uint32(len(data))})
+	if err == nil {
+		copy(data, resp.Data)
+		return len(resp.Data), nil
+	}
+
+	// check if it is &grpc.rpcError{code:0xb, desc:"EOF"} and return io.EOF instead
+	if grpc.Code(err) == codes.OutOfRange && grpc.ErrorDesc(err) == "EOF" {
+		return 0, io.EOF
+	}
+	return 0, err
+}
+
+type errPipe struct {
+	ctx         context.Context
+	agent       *shimAgent
+	containerId string
+	pid         uint32
+}
+
+func (p *errPipe) Read(data []byte) (n int, err error) {
+	resp, err := p.agent.ReadStderr(p.ctx, &pb.ReadStreamRequest{
+		ContainerId: p.containerId,
+		PID:         p.pid,
+		Len:         uint32(len(data))})
+	if err == nil {
+		copy(data, resp.Data)
+		return len(resp.Data), nil
+	}
+
+	// check if it is &grpc.rpcError{code:0xb, desc:"EOF"} and return io.EOF instead
+	if grpc.Code(err) == codes.OutOfRange && grpc.ErrorDesc(err) == "EOF" {
+		return 0, io.EOF
+	}
+	return 0, err
+}
+
```

Comment:
I would factorise the *out code and define:

```Go
type outPipe struct {
	ctx         context.Context
	agent       *shimAgent
	containerId string
	pid         uint32
        bool        errPipe
}
```

And then:

```Go
func (p *outPipe) Read(data []byte) (n int, err error) {
        var readFn func(ctx context.Context, in *ReadStreamRequest, opts ...grpc1.CallOption) (*ReadStreamResponse, error)

         if p.errPipe {
               readFn = p.agent.ReadStderr
         } else {
               readFn = p.agent.ReadStdout
         }

	resp, err := readFn(p.ctx, &pb.ReadStreamRequest{
		ContainerId: p.containerId,
		PID:         p.pid,
		Len:         uint32(len(data))})
	if err == nil {
		copy(data, resp.Data)
		return len(resp.Data), nil
	}

	// check if it is &grpc.rpcError{code:0xb, desc:"EOF"} and return io.EOF instead
	if grpc.Code(err) == codes.OutOfRange && grpc.ErrorDesc(err) == "EOF" {
		return 0, io.EOF
	}
	return 0, err
}
```

Reply:
@sameo Good point! I'm keeping the pipe struct in sync but using a helper function for both outPipe and errPipe Read() functions. Please see if the new code looks good to you.